src/cli/java/org/commcare/util/screen/MultiSelectEntityScreen.java

```diff
@@ -0,0 +1,117 @@
+package org.commcare.util.screen;
+
+import static org.commcare.session.SessionFrame.STATE_MULTIPLE_DATUM_VAL;
+
+import org.commcare.core.interfaces.VirtualDataInstanceCache;
+import org.commcare.data.xml.VirtualInstances;
+import org.commcare.modern.session.SessionWrapper;
+import org.commcare.session.CommCareSession;
+import org.commcare.suite.model.MultiSelectEntityDatum;
+import org.javarosa.core.model.instance.TreeReference;
+import org.javarosa.core.model.instance.VirtualDataInstance;
+
+import java.util.UUID;
+
+import javax.annotation.Nullable;
+
+/**
+ * Variation of EntityScreen to allow for selection of multiple entities at once
+ */
+public class MultiSelectEntityScreen extends EntityScreen {
+
+    public static final String USE_SELECTED_VALUES = "use_selected_values";
+    private int maxSelectValue = -1;
+
+    private final VirtualDataInstanceCache virtualDataInstanceCache;
+    private UUID storageReferenceId;
+
+    public MultiSelectEntityScreen(boolean handleCaseIndex, boolean full,
+            SessionWrapper session,
+            VirtualDataInstanceCache virtualDataInstanceCache)
+            throws CommCareSessionException {
+        super(handleCaseIndex, full, session);
+        this.virtualDataInstanceCache = virtualDataInstanceCache;
+    }
+
+    /**
+     * Updates entities selection on the screen
+     *
+     * @param input          Either 'use_selected_values' or a guid unique to the selection made
+     * @param selectedValues Ids of the entities selected on the screen
+     * @throws CommCareSessionException
+     */
+    @Override
+    public void updateSelection(String input, @Nullable String[] selectedValues) throws CommCareSessionException {
+        setSelectedEntities(input, selectedValues);
+    }
+
+    private void setSelectedEntities(String input, @Nullable String[] selectedValues)
+            throws CommCareSessionException {
+        if (input.contentEquals(USE_SELECTED_VALUES)) {
+            processSelectedValues(selectedValues);
+        } else {
+            UUID inputId = UUID.fromString(input);
```

Comment:
I'm fairly confused about what this clause and the `USE_SELECTED_VALUES` structure does exactly It might be worth refactoring into separate methods whcih more clearly articulate why the same call path has very different behaviors.

Reply:
will do, there are some details on this behaviour [here](https://github.com/dimagi/formplayer/wiki/Multi-Select-Case-Lists#http-lifecycle-bw-web-apps-and-formplayer)